rfd/0086-stream-events-api.md

```diff
@@ -0,0 +1,234 @@
+---
+authors: Vitor Enes (vitor@goteleport.com)
+state: draft
+---
+
+# RFD 86 - Stream Events API
+
+## Required Approvals
+
+* Engineering: @r0mant && @jimbishopp
+
+## Table of Contents
+
+* [What](#what)
+* [Why](#why)
+  * [Goals](#goals)
+  * [Non\-Goals](#non-goals)
+* [Details](#details)
+  * [`Client.StreamEvents` API](#clientstreamevents-api)
+  * [`StreamEvents` RPC](#streamevents-rpc)
+  * [`IAuditLog.StreamEvents` API](#iauditlogstreamevents-api)
+  * [`dynamoevents.Log.StreamEvents` API](#dynamoeventslogstreamevents-api)
+	* [DynamoDB stream cursor](#dynamodb-stream-cursor)
+	* [`lib/backend/dynamo/shards.go`](#libbackenddynamoshardsgo)
+* [Concerns and open questions](#concerns-and-open-questions)
+
+## What
+
+Currently streaming of audit events has to be implemented on top of [`SearchEvents`], as [done by the event-handler] plugin.
+However, this may be implemented more efficiently, specially for databases with [streaming capabilities such as DynamoDB].
+
+This RFD proposes a way to extend the [Teleport Client] with a new `StreamEvents` API that will be initially implemented only for the DynamoDB audit log.
+
+[`SearchEvents`]: https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/api/client/client.go#L2137
+[done by the event-handler]: https://github.com/gravitational/teleport-plugins/blob/142b8bd7127fe680d42fc4d626d18b3fd8df2b49/event-handler/teleport_events_watcher.go#L175-L184
+[streaming capabilities such as DynamoDB]: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Streams.html
+[Teleport Client]: https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/api/client/client.go
+
+## Why
+
+The Cloud team wants to start tracking the number of monthly active users (MAU) in order to understand the usage and growth of Teleport Cloud.
+To achieve this, audit events will be anonymized and sent to some time-series database (TBD).
+For efficiency, the plan is to implement this by leveraging DynamoDB streaming capabilities.
+
+### Goals
+
+* Add a new `StreamEvents` API to the Teleport Client
```

Comment:
Yes, that's what I meant. We don't strictly need `StreamEvents` in the `api` module. Even though it could be useful to some OSS users, it would only work for Dynamo (as you noted [here](https://github.com/gravitational/teleport/pull/16235#discussion_r966039670)), so it might make sense to instead come up with some internal API.

(Another idea was to eventually refactor the `event-handler` to use `StreamEvents` (for performance) but now I realize that, since `StreamEvents` would only for Dynamo, we would either lose functionality or have to maintain the two implementations (with `SearchEvents` and `StreamEvents`)).